cpp/src/fil/infer.cu

```diff
@@ -88,59 +214,117 @@ __global__ void infer_k(storage_type forest, predict_params params) {
-    infer_one_tree<NITEMS>(forest[j], sdata, params.num_cols, out);
-  }
-  using BlockReduce = cub::BlockReduce<vec<NITEMS>, FIL_TPB>;
-  __shared__ typename BlockReduce::TempStorage tmp_storage;
-  out = BlockReduce(tmp_storage).Sum(out);
-  if (threadIdx.x == 0) {
-    for (int i = 0; i < NITEMS; ++i) {
-      int row = blockIdx.x * NITEMS + i;
-      if (row < params.num_rows)
-        params.preds[row * params.num_output_classes] = out[i];
-    }
+    acc.accumulate(infer_one_tree<NITEMS, leaf_output_t<leaf_payload_type>::T>(
+      forest[j], sdata, params.num_cols));
   }
+  acc.finalize(params.preds, params.num_rows, params.num_outputs);
 }
 
-template <typename storage_type>
-void infer(storage_type forest, predict_params params, cudaStream_t stream) {
+template <int NITEMS, leaf_value_t leaf_payload_type>
+size_t get_smem_footprint(predict_params params) {
+  size_t finalize_footprint =
+    tree_aggregator_t<NITEMS, leaf_payload_type>::smem_finalize_footprint(
+      params.num_classes);
+  size_t accumulate_footprint =
+    sizeof(float) * params.num_cols * NITEMS +
+    tree_aggregator_t<NITEMS, leaf_payload_type>::smem_accumulate_footprint(
+      params.num_classes);
+
+  return std::max(accumulate_footprint, finalize_footprint);
+}
+
+template <leaf_value_t leaf_payload_type, typename storage_type>
+void infer_k_launcher(storage_type forest, predict_params params,
+                      cudaStream_t stream) {
   const int MAX_BATCH_ITEMS = 4;
   params.max_items =
     params.algo == algo_t::BATCH_TREE_REORG ? MAX_BATCH_ITEMS : 1;
-  int num_items = params.max_shm / (sizeof(float) * params.num_cols);
+
+  int num_items = 0;
+  size_t shm_sz = 0;
+  // solving this linear programming problem in a single equation
+  // looks less tractable than this
+  for (int nitems = 1; nitems <= params.max_items; ++nitems) {
+    size_t peak_footprint;
+    switch (nitems) {
+      case 1:
+        peak_footprint = get_smem_footprint<1, leaf_payload_type>(params);
+        break;
+      case 2:
+        peak_footprint = get_smem_footprint<2, leaf_payload_type>(params);
+        break;
+      case 3:
+        peak_footprint = get_smem_footprint<3, leaf_payload_type>(params);
+        break;
+      case 4:
+        peak_footprint = get_smem_footprint<4, leaf_payload_type>(params);
+        break;
+      default:
+        ASSERT(false, "internal error: nitems > 4");
+    }
+    // for data row
+    if (peak_footprint <= params.max_shm) {
+      num_items = nitems;
+      shm_sz = peak_footprint;
+    }
+  }
   if (num_items == 0) {
-    int max_cols = params.max_shm / sizeof(float);
-    ASSERT(false, "p.num_cols == %d: too many features, only %d allowed",
-           params.num_cols, max_cols);
+    int real_num_cols = params.num_cols;
+    // since we're crashing, this will not take too long
+    while (get_smem_footprint<1, leaf_payload_type>(params) > params.max_shm)
+      --params.num_cols;
```

Comment:
`params.num_cols` can be very big, so it can take too long. However, if you start with the number of columns  `params.max_shm / sizeof(float)`, this should be faster.